### ISSUE TITLE
Cannot run executable in Linux job due to bad file permissions set by Nomad

### ISSUE THREAD

**Issue by marcofiocco** [hashicorp/nomad#6662]
### Nomad version
0.10.1

### Operating system and Environment details
Linux Centos7

### Issue
I have an exec nomad job that downloads an artifact and unzips it, but the unzipped executable has 700 permissions and it's owned by root, while the original was 755.
I have manually tested that the unzipped file preserves 755 but Nomad apparently is changing the permissions to 700.
As a result the command line to start the executable fails with "Permission denied".
In the command line of the job I cannot do any 'chmod -R +rx' because Nomad apparently is not root and 'sudo' is not allowed by Nomad.
The following seemed to work previously, but now it does not work anymore:
```
artifact {
       source = "MYURL/myapp.zip"
} 
config {
        args = ["-c", "cd local && exec ./myapp"]
        command = "/bin/sh"
}
```
has Nomad changed behaviour in a recent version?

**Comment by tgross** [comment]
Hi @marcofiocco! I saw this same issue on the mailing list and replied to it. Would you rather continue the discussion there or move it here?

**Comment by marcofiocco** [comment]
Maybe here is tracked better

**Comment by tgross** [comment]
Ok, copying all that over...


Henry Vindin wrote:
> Looks an awful lot like a umask'ing issue. Might be worse having a quick poke around the artifact handling code for anything umasky as well as checking if the default umask on the target machine or within the container has changed. I’m not sure if the file is unzipped locally (and I assume therefore potentially impacted by a umask setting for the nomad client process, or if it’s unzipped within the container and therefore impacted by container defaults.

@tgross wrote:
> I suspect Henry is on the right track with umasks. The decompression step of our go-getter library https://github.com/hashicorp/go-getter/blob/master/decompress_zip.go doesn't mess with the permissions of the underlying files. There was a security fix in 0.9.5 to protect against privilege escalation via setuid binaries: https://github.com/hashicorp/nomad/blob/master/CHANGELOG.md#095-21-august-2019 but you're probably not running setuid application. 
>
> You say that Nomad isn't root? Typically you'll need the Nomad agent to be running as root so that it can set up the exec environment

**Comment by tgross** [comment]
@marcofiocco I'm looking at the go-getter code, in particular [`decompress_zip.go#L81-L97`](https://github.com/hashicorp/go-getter/blob/d330e2ade1c676430b969e1bd103e1171e56cec7/decompress_zip.go#L81-L97). It's creating the file and writing it, and then setting the mode to the mode of the archived file. But if there's an error at any point, it's possible to get the file created (or _partially_ created) and then fail before we set the mode.

Two things to check:
- does the SHA of the created file match what you'd expect?
- are there any error messages on the Nomad client?
